experiments/ClimaCore/sea-ice/slab_ice.jl

```diff
@@ -0,0 +1,341 @@
+"""
+function ∑tendencies_atm!(du, u, p, t)
+
+    FT = eltype(u)
+    T = u.T # u.x = vector of prognostic variables from DifferentialEquations
+    F_sfc = calculate_flux(p.Ya.T_sfc[1], parent(T)[1], p.p)
+
+    ## set BCs
+    bcs_bottom = Operators.SetValue(Geometry.WVector(F_sfc ./ p.p.λ))
+    bcs_top = Operators.SetValue(Geometry.WVector(FT(0)))
+
+    gradc2f = Operators.GradientC2F()
+    gradf2c = Operators.DivergenceF2C(bottom = bcs_bottom, top = bcs_top) # Neumann BC (face-to-center)
+
+    ## tendency calculations
+    @. du.T = gradf2c(p.p.μ * gradc2f(T)) # dT/dt
+    du.F .= -F_sfc[1] # d(F_integrated)/dt
+
+end
+
+
+get_∂F_atm∂T_sfc(p) = -p.λ # TODO: in GCM this will be: 4 σ T_sfc^3 + ||uh|| ρ cp (C_H + L(q_sat1 - q_sat) * dT^(-1))
+
+# - Model 2 (ice) Equations
+"""
+    solve_ice!(dT_sfc, T_sfc, (parameters, F_accumulated), t)
+
+slab RHS with an implicit solve ice and explicit (forward Euler) solve for ocean
+
+"""
+function solve_ice!(integ, Δt)
+
+    Y = integ.u
+    Ya = integ.p.Ya
+    p = integ.p.p
+
+    FT = eltype(Y)
+    ocean_qflux = FT(0)
+
+    # prognostic
+    T_ml = Y.T_ml
+    T_sfc = Y.T_sfc
+    h_ice = Y.h_ice
+
+    # auxiliary
+    F_atm = Ya.F_atm
+
+    ∂F_atm∂T_sfc = get_∂F_atm∂T_sfc(p) # this will be passed from atmos/SF.jl
+
+    ΔT_ml = similar(F_atm)
+    Δh_ice = similar(F_atm)
+    F_ice = similar(F_atm)
+    ΔT_base = similar(F_atm)
+
+    # ice thickness and mixed layer temperature changes due to atmosphereic and ocean fluxes 
+    if h_ice[1] <= 0 # ice-free 
+        @. ΔT_ml = -(F_atm + ocean_qflux) * Δt / (p.h_ml * p.ρc_ml)
+        @. Δh_ice = 0
+    else # ice-covered
+        @. ΔT_ml = -(p.F_basal * (T_ml - p.T_base) + ocean_qflux) * Δt / (p.h_ml * p.ρc_ml)
+        @. Δh_ice = (F_atm - p.F_basal * (T_ml - p.T_base)) * Δt / p.L_ice
+    end
+
+    # adjust if transition to ice-covered 
+    if (T_ml[1] + ΔT_ml[1] < p.T_freeze)
+        @. Δh_ice = Δh_ice - (T_ml + ΔT_ml - p.T_freeze) * (p.h_ml * p.ρc_ml) / p.L_ice
+        @. ΔT_ml = p.T_freeze - T_ml
+    end
+
+    # adjust if transition to ice-free
+    if ((h_ice[1] > 0) & (h_ice[1] + Δh_ice[1] <= 0))
+        @. ΔT_ml = ΔT_ml - (h_ice + Δh_ice) * p.L_ice / (p.h_ml * p.ρc_ml)
+        @. Δh_ice = -h_ice
+    end
+
+    # solve for T_sfc
+    if (h_ice[1] + Δh_ice[1] > 0) #  surface is ice-covered
+        # if ice covered, solve implicity (for now one Newton iteration: ΔT_s = - F(T_s) / dF(T_s)/dT_s )
+        @. F_ice = p.k_ice / (h_ice + Δh_ice) * (p.T_base - T_sfc)
+        @. ΔT_base = (-F_atm + F_ice) / (p.k_ice / (h_ice + Δh_ice) + ∂F_atm∂T_sfc)
+        if (T_sfc[1] + ΔT_base[1] > p.T_freeze)
+            @. ΔT_base = p.T_freeze - T_sfc
+        end
+        # surface is ice-covered, so update T_sfc as ice surface temperature
+        @. Y.T_sfc += ΔT_base
+    else # ice-free, so update T_sfc as mixed layer temperature
+        @. Y.T_sfc = T_ml + ΔT_ml
+    end
+
+    # update state
+    @. Y.T_ml += ΔT_ml
```

Comment:
Does this work? I thought you could only change prognostic variables via callbacks (or at least, that it was only recommended to do so). 

Would it be possibly to implement the ice model as a diff eq now, rather than implement this kind of funky "set Y by timestepping by hand, wrap inside a ODE problem which doesnt ever change dY from 0" (also, is dY *always* 0 when initialized? perhaps you should still set dY .= zero field inside the ice tendencies stub)

Reply:
Yeah, it works lol. I did it this way mainly to deal with the time stepping and to be able to access the ice state from previous time steps for diagnostics. We can do something more sophisticated in the next iteration, but it may not be straightforward since you need to do stuff between time integrating `h_ice` and `dT_sfc`.